File: diff-apidocs.mjs
#!/usr/bin/env node
/**
 * diff-apidocs.mjs — diff the public API surface between two archived product versions
 * (Node port of diff-apidocs.py; output format matches it exactly).
 *
 * Reads only the committed object-reference archive; nothing is downloaded and no build is
 * triggered. Identity comes from each version's xrefmap.yml, signatures from its DocFX api/*.html
 * pages.
 *
 *   node diff-apidocs.mjs -p ironzip
 *   node diff-apidocs.mjs -p ironzip --from 2026.5.2 --to 2026.6.2
 *   node diff-apidocs.mjs -p ironpdf --namespace 'IronPdf.Rendering.*' --markdown
 *   node diff-apidocs.mjs -p ironzip --json --fail-on-breaking
 *
 * Exit codes: 0 success, 1 tool error (unknown product, missing version), 2 breaking changes found
 * with --fail-on-breaking.
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, isAbsolute, join, resolve } from "node:path";
import { parseArgs } from "node:util";
import { fileURLToPath } from "node:url";

import { listArchivedVersions } from "./apidocs.mjs";
import { StatusLogger } from "./statuslogger.mjs";
import {
  ArchiveError, loadProduct, requireSupported, resolveVersions, versionDirectory,
} from "./apidiff/archive.mjs";
import { bySeverity, diffSurfaces } from "./apidiff/classify.mjs";
import { SurfaceFilter } from "./apidiff/filters.mjs";
import { BREAKING } from "./apidiff/model.mjs";
import { render as renderJson } from "./apidiff/render-json.mjs";
import { render as renderMarkdown } from "./apidiff/render-markdown.mjs";
import { render as renderText } from "./apidiff/render-text.mjs";
import { buildSurface } from "./apidiff/surface.mjs";

const CWD = dirname(fileURLToPath(import.meta.url));

/**
 * Committed diff artifacts live under docs/, which _config.yml already excludes from the Jekyll
 * build, so they stay private to the repo.
 */
const ARTIFACT_ROOT = join(CWD, "docs", "api-diffs");

const EXIT_OK = 0;
const EXIT_ERROR = 1;
const EXIT_BREAKING = 2;

const USAGE = `usage: diff-apidocs [-h] [-p PRODUCT_CODE] [-n PRODUCT_NAME] [--from VERSION] [--to VERSION]
                    [--namespace GLOB] [--exclude GLOB] [--include-internal] [--all-visibility]
                    [--json [PATH]] [--markdown [PATH]] [--quiet] [--no-warnings]
                    [--fail-on-breaking] [--list-versions]

Diff the public API surface between two archived versions of an Iron Software product.

  -p, --product-code    Product short code, e.g. ironzip
  -n, --product-name    Product display name, e.g. IronZIP
      --from            Older version (default: previous archived)
      --to              Newer version (default: newest archived)
      --namespace       Only report types matching this glob; repeatable
      --exclude         Skip types matching this glob; repeatable
      --include-internal  Include vendored/internal namespaces (excluded by default)
      --all-visibility  Include non-public declarations (public/protected only by default)
      --json [PATH]     Write JSON; defaults to docs/api-diffs/<code>/<from>..<to>.json
      --markdown [PATH] Write Markdown; defaults alongside the JSON artifact
      --quiet           Suppress the terminal report
      --no-warnings     Suppress parser warnings
      --fail-on-breaking  Exit 2 when breaking changes are found
      --list-versions   List the product's archived versions and exit`;

const OPTIONS = {
  help: { type: "boolean", short: "h", default: false },
  "product-code": { type: "string", short: "p" },
  "product-name": { type: "string", short: "n" },
  from: { type: "string" },
  to: { type: "string" },
  namespace: { type: "string", multiple: true, default: [] },
  exclude: { type: "string", multiple: true, default: [] },
  "include-internal": { type: "boolean", default: false },
  "all-visibility": { type: "boolean", default: false },
  // parseArgs has no optional-value form, so these are strings defaulting to "" when bare.
  json: { type: "string" },
  markdown: { type: "string" },
  quiet: { type: "boolean", default: false },
  "no-warnings": { type: "boolean", default: false },
  "fail-on-breaking": { type: "boolean", default: false },
  "list-versions": { type: "boolean", default: false },
};

/**
 * Allow `--json` and `--markdown` to be passed with no value, matching argparse's `nargs="?"`.
 * A bare flag becomes "" (use the default artifact path); anything else is taken as a path.
 */
function normalizeOptionalValueFlags(argv) {
  const normalized = [];
  for (let index = 0; index < argv.length; index++) {
    const argument = argv[index];
    if (argument === "--json" || argument === "--markdown") {
      const next = argv[index + 1];
      if (next === undefined || next.startsWith("-")) {
        normalized.push(`${argument}=`);
        continue;
      }
    }
    normalized.push(argument);
  }
  return normalized;
}

function artifactPath(supplied, productCode, versionFrom, versionTo, extension) {
  if (supplied) return isAbsolute(supplied) ? supplied : resolve(supplied);
  return join(ARTIFACT_ROOT, productCode, `${versionFrom}..${versionTo}.${extension}`);
}

function write(path, content) {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content, "utf-8");
  StatusLogger.progress(`Wrote ${path}`);
}

export function main(argv) {
  let values;
  try {
    ({ values } = parseArgs({ args: normalizeOptionalValueFlags(argv), options: OPTIONS, allowPositionals: false }));
  } catch (error) {
    StatusLogger.error(error.message);
    console.log(USAGE);
    return EXIT_ERROR;
  }

  if (values.help) {
    console.log(USAGE);
    return EXIT_OK;
  }

  let product;
  let versionFrom;
  let versionTo;
  let directoryFrom;
  let directoryTo;
  try {
    product = loadProduct(values["product-code"] ?? null, values["product-name"] ?? null);
    requireSupported(product);

    if (values["list-versions"]) {
      const versions = listArchivedVersions(product.code);
      StatusLogger.title(`${product.name} (${product.code}) — ${versions.length} archived versions`);
      for (const version of versions) StatusLogger.message(`  ${version}`);
      return EXIT_OK;
    }

    [versionFrom, versionTo] = resolveVersions(product, values.from ?? null, values.to ?? null);
    directoryFrom = versionDirectory(product, versionFrom);
    directoryTo = versionDirectory(product, versionTo);
  } catch (error) {
    if (!(error instanceof ArchiveError)) throw error;
    StatusLogger.error(error.message);
    return EXIT_ERROR;
  }

  const surfaceFilter = new SurfaceFilter({
    includeInternal: values["include-internal"],
    namespaces: values.namespace,
    excludes: values.exclude,
    publicOnly: !values["all-visibility"],
  });

  const surfaceFrom = buildSurface(directoryFrom, product.code, versionFrom, surfaceFilter);
  const surfaceTo = buildSurface(directoryTo, product.code, versionTo, surfaceFilter);
  const result = diffSurfaces(surfaceFrom, surfaceTo, product.name);

  if (!values.quiet) renderText(result, !values["no-warnings"]);

  if (values.json !== undefined) {
    write(artifactPath(values.json, product.code, versionFrom, versionTo, "json"), renderJson(result));
  }
  if (values.markdown !== undefined) {
    write(artifactPath(values.markdown, product.code, versionFrom, versionTo, "md"), renderMarkdown(result));
  }

  if (values["fail-on-breaking"] && bySeverity(result, BREAKING).length) return EXIT_BREAKING;
  return EXIT_OK;
}

process.exitCode = main(process.argv.slice(2));

File: diff-apidocs.py
#!/usr/bin/env python3
"""diff-apidocs.py — Diff the public API surface between two archived product versions.

Reads only the committed object-reference archive; nothing is downloaded and no build is triggered.
Identity comes from each version's xrefmap.yml, signatures from its DocFX api/*.html pages.

Examples:
    python diff-apidocs.py -p ironzip
    python diff-apidocs.py -p ironzip --from 2026.5.2 --to 2026.6.2
    python diff-apidocs.py -p ironpdf --namespace 'IronPdf.Rendering.*' --markdown
    python diff-apidocs.py -p ironzip --json --fail-on-breaking

Exit codes: 0 success, 1 tool error (unknown product, missing version), 2 breaking changes found
with --fail-on-breaking.
"""

import argparse
import os
import sys

# Some archived uids carry Unicode Private Use Area markers that DocFX emits for unresolvable type
# parameters (U+E000, U+E396 and U+E397 appear in ironpdf 2023.11.7, for example). Writing those to a
# legacy cp1252 Windows console raises UnicodeEncodeError and kills the run, so force UTF-8 here.
# Scoped to this entry point so the shared StatusLogger keeps its existing behaviour for other tools.
for _stream in (sys.stdout, sys.stderr):
    if hasattr(_stream, "reconfigure"):
        _stream.reconfigure(encoding="utf-8")

from apidiff.archive import (ArchiveError, load_product, require_supported, resolve_versions,
                             version_directory)
from apidiff.classify import diff_surfaces
from apidiff.filters import SurfaceFilter
from apidiff.model import BREAKING
from apidiff.render_json import render as render_json
from apidiff.render_markdown import render as render_markdown
from apidiff.render_text import render as render_text
from apidiff.surface import build_surface
from statuslogger import StatusLogger

# Committed diff artifacts live under docs/, which _config.yml already excludes from the Jekyll
# build, so they stay private to the repo.
ARTIFACT_ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "docs", "api-diffs")

EXIT_OK = 0
EXIT_ERROR = 1
EXIT_BREAKING = 2


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="diff-apidocs",
        description="Diff the public API surface between two archived versions of an Iron Software product.",
    )
    parser.add_argument("-p", "--product-code", help="Product short code, e.g. ironzip")
    parser.add_argument("-n", "--product-name", help="Product display name, e.g. IronZIP")
    parser.add_argument("--from", dest="version_from", help="Older version (default: previous archived)")
    parser.add_argument("--to", dest="version_to", help="Newer version (default: newest archived)")
    parser.add_argument("--namespace", action="append", default=[],
                        help="Only report types matching this glob; repeatable")
    parser.add_argument("--exclude", action="append", default=[],
                        help="Skip types matching this glob; repeatable")
    parser.add_argument("--include-internal", action="store_true",
                        help="Include vendored/internal namespaces (excluded by default)")
    parser.add_argument("--all-visibility", action="store_true",
                        help="Include non-public declarations (public/protected only by default)")
    parser.add_argument("--json", nargs="?", const="", metavar="PATH",
                        help=f"Write JSON; defaults to {os.path.join('docs', 'api-diffs')}/<code>/<from>..<to>.json")
    parser.add_argument("--markdown", nargs="?", const="", metavar="PATH",
                        help="Write Markdown; defaults alongside the JSON artifact")
    parser.add_argument("--quiet", action="store_true", help="Suppress the terminal report")
    parser.add_argument("--no-warnings", action="store_true", help="Suppress parser warnings")
    parser.add_argument("--fail-on-breaking", action="store_true",
                        help="Exit 2 when breaking changes are found")
    parser.add_argument("--list-versions", action="store_true",
                        help="List the product's archived versions and exit")
    return parser


def _artifact_path(supplied: str, product_code: str, version_from: str, version_to: str, extension: str) -> str:
    if supplied:
        return os.path.abspath(supplied)
    return os.path.join(ARTIFACT_ROOT, product_code, f"{version_from}..{version_to}.{extension}")


def _write(path: str, content: str) -> None:
    os.makedirs(os.path.dirname(path), exist_ok=True)
    # newline="" keeps the "\n" line endings the renderers emit, so the two ports stay byte-identical
    # on Windows as well as Linux.
    with open(path, "w", encoding="utf-8", newline="") as handle:
        handle.write(content)
    StatusLogger.progress(f"Wrote {path}")


def main(argv) -> int:
    args = build_parser().parse_args(argv)

    try:
        product = load_product(args.product_code, args.product_name)
        require_supported(product)

        if args.list_versions:
            from apidocs import list_archived_versions
            versions = list_archived_versions(product["code"])
            StatusLogger.title(f"{product['name']} ({product['code']}) — {len(versions)} archived versions")
            for version in versions:
                StatusLogger.message(f"  {version}")
            return EXIT_OK

        version_from, version_to, _ = resolve_versions(product, args.version_from, args.version_to)
        directory_from = version_directory(product, version_from)
        directory_to = version_directory(product, version_to)
    except ArchiveError as error:
        StatusLogger.error(str(error))
        return EXIT_ERROR

    surface_filter = SurfaceFilter(
        include_internal=args.include_internal,
        namespaces=args.namespace,
        excludes=args.exclude,
        public_only=not args.all_visibility,
    )

    surface_from = build_surface(directory_from, product["code"], version_from, surface_filter)
    surface_to = build_surface(directory_to, product["code"], version_to, surface_filter)
    result = diff_surfaces(surface_from, surface_to, product["name"])

    if not args.quiet:
        render_text(result, show_warnings=not args.no_warnings)

    if args.json is not None:
        _write(_artifact_path(args.json, product["code"], version_from, version_to, "json"), render_json(result))
    if args.markdown is not None:
        _write(_artifact_path(args.markdown, product["code"], version_from, version_to, "md"), render_markdown(result))

    if args.fail_on_breaking and result.by_severity(BREAKING):
        return EXIT_BREAKING
    return EXIT_OK


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
